packages/yarn-plugin-allow-scripts/test/index.spec.mts

```diff
@@ -0,0 +1,238 @@
+  try {
+    const defaultOpts = realisticEnvOptions(cwd);
+    const result = spawnSync(
+      isWindows
+        ? process.env.COMSPEC || 'cmd.exe'
+        : process.env.SHELL || '/bin/sh',
+      [
+        ...(isWindows ? ['/d', '/s', '/c'] : ['-c']),
+        [
+          isWindows ? 'corepack.cmd' : 'corepack',
+          'yarn@4',
+          ...args
+        ].join(' '),
+      ],
+      {
+        ...defaultOpts,
+        env: {
+          ...process.env,
+          ...defaultOpts.env,
+          ...env,
+        }
+      },
+    )
+    return {
+      exitCode: result.status,
+      out: result?.stdout?.toString()?.trim(),
+      err: result?.stderr?.toString()?.trim(),
+      output: result?.output,
+      signal: result.signal,
+      error: result.error,
+    }
+  } catch(error) {
+    console.error(error);
+    t.fail(
+      `Failed calling '${args.join(' ')}': ${JSON.stringify(
+        {
+          cwd,
+          error,
+        },
+        undefined,
+        2
+      )}`
+    )
+    return {
+      error,
+      output: [],
+      exitCode: null,
+      signal: null,
+      out: '',
+      err: '',
+    };
+  }
+}
+
+test('install - errors when adding dependency with unspecified preinstall script', async (t: any) => {
+  const projectName = 'uninitialized';
+  const projectRoot = join(dirname(import.meta.url.replace(/^file:/, '')), 'projects', projectName);
+
+  try {
+    // delete any leftover test artifacts
+    rmSync(join(projectRoot, PACKAGE_JSON), { force: true })
+
+    // init project
+    const initRes = run(t, ['init', '-y'], projectRoot)
+    t.is(initRes.exitCode, 0);
+
+    // install the plugin
+    const importRes = run(t, ['plugin', 'import', normalize('../../../bundles/@yarnpkg/plugin-allow-scripts.js')], projectRoot)
+    t.is(importRes.err, '');
+    t.regex(importRes.out, /YN0000: Saving the new plugin in .yarn\/plugins\/@yarnpkg\/plugin-allow-scripts.cjs/);
+    t.is(importRes.exitCode, 0);
+
+    const addRes = run(t, ['add', '@lavamoat/preinstall-always-fail'], projectRoot)
+    t.regex(addRes.err, /allow-scripts detected attempted execution of unconfigured package script. \["@lavamoat\/preinstall-always-fail","preinstall"/);
+    t.is(addRes.exitCode, 1);
+
+    // trigger the plugin
+    const installRes = run(t, ['install'], projectRoot)
+    t.regex(installRes.out, /packages missing configuration:\n- @lavamoat\/preinstall-always-fail/);
+    t.regex(installRes.err, /@lavamoat\/allow-scripts has detected dependencies without configuration. explicit configuration required./);
+    t.is(installRes.exitCode, 1);
+  } finally {
+    cleanup(projectName);
+  }
+})
+
+test('install - allows adding and installing package with specified preinstall script', async (t: any) => {
+  const projectName = 'allowed';
+  const projectRoot = join(dirname(import.meta.url.replace(/^file:/, '')), 'projects', projectName)
```

Comment:
(could) extract to a utility function or maybe even hide inside `run`
(could) slight preference for calling the folder fixtures not projects, but up to you.

Reply:
> (could) extract to a utility function or maybe even hide inside run
Seems premature until we see further usage

> (could) slight preference for calling the folder fixtures not projects, but up to you.
In general agreed but following existing norm from `allow-scripts`. I'd rather rename both in a follow-up.

Reply:
> (could) extract to a utility function

Done